### PR TITLE
EZP-25206: Display basic trash in PlatformUI

### PR DESCRIPTION
**Pull request by yannickroger**
Link: https://jira.ez.no/browse/EZP-25206

## Description
The PR introduces the first version of the Trash. It adds a button in the menu on the left and allows displaying items located in the Trash.

Note: This code highlights this issue: https://jira.ez.no/browse/EZP-25404

## Screenshot
trash

## Tests
Manual and unit tests

## TODO
- [ ] Create an improvement to have original location in trash item.

### CODE REVIEW

File: Resources/config/yui.yml [hunk 1]
```diff
@@ -173,6 +175,19 @@ system:
...
                     path: %ez_platformui.public_dir%/js/views/services/ez-contentcreateviewservice.js
+                ez-trashviewservice:
+                    requires:
+                        - 'parallel'
+                        - 'ez-viewservice'
+                        - 'ez-trashitemmodel'
+                        - 'ez-contentinfomodel'
+                        - 'ez-locationmodel'
+                        - 'ez-contenttypemodel'
+                        - 'ez-usermodel'
```
Comment (dpobel) [inline Resources/config/yui.yml]: this and the following 2 modules can be removed from the dependencies.

File: Resources/config/yui.yml [hunk 2]
```diff
@@ -445,6 +460,15 @@ system:
                 contenteditview-ez-template:
                     type: 'template'
                     path: %ez_platformui.public_dir%/templates/contentedit.hbt
+                ez-trashview:
+                    requires:
+                        - 'ez-templatebasedview'
+                        - 'event-tap'
```
Comment (dpobel) [inline Resources/config/yui.yml]: useless `event-tap`

File: Resources/public/js/views/ez-trashview.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+         *
+         * @protected
+         * @method _convertTrashItemsToJSON
+         * @return {Object} JSONified `trashItems`
+         */
+        _convertTrashItemsToJSON: function () {
+            var trashItemsJSON = [],
+                trashItemJSON;
+
+            Y.Array.each(this.get('trashItems'), function (trashItem) {
```
Comment (dpobel) [inline Resources/public/js/views/ez-trashview.js]: can be written in a bit more elegant way with `Y.Array.map`

**Comment by yannickroger** [comment]
Ready for review @dpobel @StephaneDiot

File: Resources/config/yui.yml [hunk 4]
```diff
@@ -788,6 +810,9 @@ system:
                 ez-locationmodel:
                     requires: ['ez-restmodel', 'ez-contentinfomodel', array-extras]
                     path: %ez_platformui.public_dir%/js/models/ez-locationmodel.js
+                ez-trashitemmodel:
+                    requires: ['ez-restmodel', array-extras]
```
Comment (dpobel) [inline Resources/config/yui.yml]: unless I'm mistaken, `array-extras` is not needed

File: Resources/public/js/views/ez-trashview.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    Y.eZ.TrashView = Y.Base.create('trashView', Y.eZ.TemplateBasedView, [], {
+
+        /**
+        * Initializer is called upon view's init
+        * Creating and managing child views inside it
+        *
+        * @method initializer
+        */
+        initializer: function () {
```
Comment (dpobel) [inline Resources/public/js/views/ez-trashview.js]: you should remove it

File: Resources/public/js/views/ez-trashview.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+            return this;
+        },
+
+        /**
+         * Converts the content of the `trashItems` attribute into JSON
+         *
+         * @protected
+         * @method _convertTrashItemsToJSON
+         * @return {Object} JSONified `trashItems`
```
Comment (dpobel) [inline Resources/public/js/views/ez-trashview.js]: nope, it returns an Array

File: Resources/public/js/views/services/ez-discoverybarviewservice.js [hunk 7]
```diff
@@ -19,5 +19,14 @@ YUI.add('ez-discoverybarviewservice', function (Y) {
...
      * @extends eZ.ViewService
      */
-    Y.eZ.DiscoveryBarViewService = Y.Base.create('discoveryBarViewService', Y.eZ.ViewService, [Y.eZ.SideViewService], {});
+    Y.eZ.DiscoveryBarViewService = Y.Base.create('discoveryBarViewService', Y.eZ.ViewService, [Y.eZ.SideViewService], {
+
+        initializer: function () {
+            this.on('*:viewTrashAction', this._redirectToTrashView);
+        },
+
+        _redirectToTrashView: function () {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-discoverybarviewservice.js]: missing doc

File: Resources/public/js/views/services/ez-trashviewservice.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+     * Loads the trash items needed by the trash view
+     *
+     * @namespace eZ
+     * @class TrashViewService
+     * @constructor
+     * @extends eZ.ViewService
+     */
+    Y.eZ.TrashViewService = Y.Base.create('trashViewService', Y.eZ.ViewService, [], {
+
+        initializer: function () {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-trashviewservice.js]: can be removed as well

File: Resources/public/js/views/services/ez-trashviewservice.js [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+            var contentService = this.get('capi').getContentService(),
+                service = this;
+
+            contentService.loadTrashItems(-1, 0, function (error, response){
+                var tasks = new Y.Parallel(),
+                    trashItems = [],
+                    trashItemsHash = response.document.Trash.TrashItem,
+                    inTrashList = service._buildInTrashList(trashItemsHash);
+
+                if(error) {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-trashviewservice.js]: nitpick: space after `if`

File: Resources/public/js/views/services/ez-trashviewservice.js [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+         *
+         * @method _load
+         * @protected
+         * @param {Function} callback
+         */
+        _load: function (callback) {
+            var contentService = this.get('capi').getContentService(),
+                service = this;
+
+            contentService.loadTrashItems(-1, 0, function (error, response){
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-trashviewservice.js]: nitpick: space before `{`

**Comment by dpobel** [comment]
besides the inline comments, good job +1

**Comment by yannickroger** [comment]
Feedback by @dpobel taken into account.

**Comment by yannickroger** [comment]
Since https://github.com/ezsystems/ezpublish-kernel/pull/1563 has been merged, I will refactor this PR to take it into account.

**Comment by yannickroger** [comment]
Code has been refactored to use the `ContentInfo` that is now present in the TrashItem.
For now, it is in a seperate commit to ease the review update ( ping @dpobel ), but I can squash it later if needed.

**Comment by dpobel** [comment]
very good :) still +1

**Comment by yannickroger** [comment]
Commits squashed, ping @StephaneDiot

**Comment by StephaneDiot** [comment]
+1

**Pull request merged**